File: floor.js
// --- checkerboard floor ---
const FLOOR_Z_NEAR = 1.0;
const FLOOR_Z_FAR = 6.0;
const FLOOR_FOCAL = 96.0;
const FLOOR_ROWS = SCREEN_H - HORIZON_Y;
const INV_Z_NEAR = 1.0 / FLOOR_Z_NEAR;
const INV_Z_FAR  = 1.0 / FLOOR_Z_FAR;

const COL_RED   = rgb565(180, 40, 40);
const COL_WHITE = rgb565(220, 220, 220);

// render perspective checkerboard floor from horizon to screen bottom
function floor_draw() {
  for (let y = HORIZON_Y - 4; y < SCREEN_H; y++) {
    let t = (SCREEN_H - 1 - y) / (FLOOR_ROWS - 1);
    let z = 1.0 / (INV_Z_NEAR + (INV_Z_FAR - INV_Z_NEAR) * t);
    let tz = (z | 0);

    for (let x = 0; x < SCREEN_W; x++) {
      let wx = (x - SCREEN_W / 2) * z / FLOOR_FOCAL;
      let tx = wx >= 0 ? (wx | 0) : ((wx | 0) - 1);

      let tile = (tx + tz) & 1;
      putpixel(x, y, tile ? COL_RED : COL_WHITE);
    }
  }
}

File: zephyr/src/floor.c
// --- checkerboard floor ---

#include "floor.h"
#include "graph.h"

#define FLOOR_Z_NEAR  1.0f
#define FLOOR_Z_FAR   6.0f
#define FLOOR_FOCAL   96.0f
#define FLOOR_ROWS    (SCREEN_H - HORIZON_Y)

static const float INV_Z_NEAR = 1.0f / FLOOR_Z_NEAR;
static const float INV_Z_FAR  = 1.0f / FLOOR_Z_FAR;

static uint16_t col_red;
static uint16_t col_white;

// pre-compute tile colors
void floor_init(void)
{
    col_red   = rgb565(180, 40, 40);
    col_white = rgb565(220, 220, 220);
}

// render perspective checkerboard floor from horizon to screen bottom
void floor_draw(void)
{
    for (int y = HORIZON_Y - 4; y < SCREEN_H; y++) {
        float t = (float)(SCREEN_H - 1 - y) / (float)(FLOOR_ROWS - 1);
        float z = 1.0f / (INV_Z_NEAR + (INV_Z_FAR - INV_Z_NEAR) * t);
        int tz = (int)z;
        int fb_row = y * SCREEN_W;

        for (int x = 0; x < SCREEN_W; x++) {
            float wx = (float)(x - SCREEN_W / 2) * z / FLOOR_FOCAL;
            int tx = (wx >= 0.0f) ? (int)wx : ((int)wx - 1);
            int tile = (tx + tz) & 1;
            fb_back[fb_row + x] = tile ? col_red : col_white;
        }
    }
}
